### PR TITLE
42 combine loggers

### PR DESCRIPTION
**Pull request by ali-i3**
# Summary

- Created a fully functioning run-logger that saves the following for each run: run_id, timestamp, version, run_time, logs, configs
- Run-logs output to 3 csv files
- Created dummy pipeline that runs

-  A prototype logging module has been created in `wrappers.py`
- Multiple dummy functions in `test_functions.py` run and use the logging module
-  The latest logging file is output to `logs/main.log` overwriting previous logs
-  Logging is output to the console depending on user preference

-:A run Id number is created from previous iterations
-  a run-logger is capturing:
      - time of run (start),
      - entire config file column separated
      - how long it took (duration)
      - all the logs from that run
-  A dummy pipeline has a config file (currently yaml) named developer_config
-  A dummy pipeline uses the logger 

- The logs are appended to a csv file
- The configs are appended to a csv file
- The pipeline run metadata is appended to a csv

These csv output files will be tidied up in RDRP-91

UPDATE:

I have since implemented the requested changes and added more doc-strings and comments.

# Checklists

This pull/merge request meets the following requirements:

- [x] code runs
- [ ] [developments are ethical][data-ethics-framework] and secure
- [ ] you have made proportionate checks that the code works correctly
- [ ] test suite passes
- [ ] developments adhere to AQA plan (see `docs/aqa/aqa_plan.md`)
- [ ] data log updated (see `docs/aqa/data_log.md`), if necessary
- [ ] assumptions, and caveats log updated (see `docs/aqa/assumptions_caveats.md`), if
  necessary
- [ ] [minimum usable documentation][agilemodeling] written in the `docs` folder

Comments have been added below around the incomplete checks.

[agilemodeling]: http://agilemodeling.com/essays/documentLate.htm
[data-ethics-framework]: https://www.gov.uk/government/publications/data-ethics-framework

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -21,7 +21,7 @@ repos:
       - id: trailing-whitespace
         name: Check for trailing whitespaces (auto-fixes)
   - repo: https://github.com/pycqa/isort
-    rev: 5.8.0
+    rev: 5.12.0
```
Comment (jwestw) [inline .pre-commit-config.yaml:24]: @ali-i3 Is this the correct version? This looks very recent (maybe the latest, before pre-release versions). Does this work with Py3.6?

Comment (jwestw) [inline .pre-commit-config.yaml:24]: @ali-i3 Is this the correct version? This looks very recent (maybe the latest, before pre-release versions). Does this work with Py3.6?

File: src/utils/testfunctions.py [hunk 2]
```diff
@@ -1,10 +1,66 @@
...
+from src.utils.wrappers import time_logger_wrap, exception_wrap
+import time
+import numpy as np
+import pandas as pd
+import logging
+
+LOGGER = logging.getLogger(__name__)
+LOGGER.setLevel(logging.INFO)
+
+@exception_wrap
```
Comment (jwestw) [inline src/utils/testfunctions.py:11]: This all looks great. No need to add anything else to the functions, because they are just for testing.

File: src/utils/wrappers.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
+import logging
+from functools import wraps
+from time import perf_counter
+import traceback
+import datetime
+
+# from table_logger import TableLogger
+import logging.config
+
+table_config = "SingleLine"
```
Comment (jwestw) [inline src/utils/wrappers.py:10]: Any config settings should go in a config file.

File: src/utils/wrappers.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+                logging.FileHandler("logs/main.log", mode='w'),
+                logging.StreamHandler(),
+            ],
+        )
+        logger = logging.getLogger(__name__)
+    else:
+        logging.basicConfig(
+            level="DEBUG", format="%(asctime)s - %(levelname)s:%(message)s"
+        )
+        logger = logging.getLogger(__name__)
```
Comment (jwestw) [inline src/utils/wrappers.py:32]: Please comment in plain English what is being tested here with the `if` `else` statements

File: src/utils/wrappers.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+#     elif table_config == "SingleLine":
+#         logger.info(
+#             f"""Difference in rows: {row_diff} ({row_change}),
+#                         Difference in columns {col_diff} ({col_change})"""
+#         )
+#     else:
+#         logger.warning(
+#             """Trouble at mill!!! Mistake in config.
+#                           Either 'Table' or 'SingleLine' must be specified."""
+#         )
```
Comment (jwestw) [inline src/utils/wrappers.py:203]: Can you remind me why this is not included and is currently commented out? It's fine if you think this part is beyond the scope of your (already massive) ticket, but please make a new ticket to implement this, plus, extract (cut) the code and put it as detail into the new ticket on Jira.

**Review by jwestw — CHANGES_REQUESTED** [review]
Really good effort. Please go through your code and action the points I have raised and also add more comments throughout the code, not on every line, but for every major step. Try to [avoid repeating yourself](https://www.geeksforgeeks.org/dry-dont-repeat-yourself-principle-in-java-with-examples/). Do this by putting some of the repetition into functions.

File: src/utils/wrappers.py (hunk 5 above)
Comment (ali-i3) [reply src/utils/wrappers.py:195]: This code was for the table logging which isn't part of ticket 42. I commented it out so the rest of the code would run (I know it isn't best practice). I've since created branch 80 as part of our sprint where I have worked on this code and it runs fine. 

I'll delete this from this branch and you can review the working code in branch-80's pull request.
